sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeArrayData.java

```diff
@@ -336,4 +336,62 @@ public UnsafeArrayData copy() {
     arrayCopy.pointTo(arrayDataCopy, Platform.BYTE_ARRAY_OFFSET, sizeInBytes);
     return arrayCopy;
   }
+
+  public int[] toPrimitiveIntArray() {
+    int[] result = new int[numElements];
+    Platform.copyMemory(baseObject, baseOffset + 4 + 4 * numElements,
```

Comment:
* `4 *` -> `4L *` to avoid overflow. Please check other places as well.
* I don't quite understand the `offsetRegionSize`. Is it reserved for marking null values or handling variable-length elements in the future? This is quite expensive for primitive arrays.